### PR TITLE
net: tcp: Fixes for issues found with loopback operation

### LINKED ISSUE #44012: net: tcp: Cooperative scheduling transfer size limited
When using a loopback to transfer data between a server and client, that is bigger then the number of available buffers, the transfer fails.

A testcase has been created in tests/net/socket/tcp on:
https://github.com/ssharks/zephyr/tree/tcp_big_data_transfer

A few observations so far:
- Send returns not for the complete transfer, but only for a chunk.
- Only settings the testcase threads to preemtive scheduling does not seem to help.

### LINKED ISSUE #44018: net: tcp: Running out of buffers by packet loss
When transferring a big TCP send, packet loss causes the send to return with a NOMEM error.

The tests/net/socket/tcp has been modified to reproduce this.
The packet loss is simulated by throwing away one every 8 packets in the loopback driver.
https://github.com/ssharks/zephyr/tree/tcp_big_data_transfer

### PR DESCRIPTION
**Pull request by rlubos**
This PR brings a set of improvements/fixes for TCP operation over loopback interface (and not only loopback for some commits):

* When cloning packet at loopback driver, use a proper packet pool
* Match loopback MTU with IPv4 MTU to prevent fragmentation
* React on zero-window sent by peer
* Improve TCP stack performance when loopback is used
* Fix a bug, where TCP stack accepted unlimited data from the application in retransmit mode

Fixes #44012
Fixes #44018

### CODE REVIEW

**Comment by rlubos** [comment]
@ssharks Please give it a shot.

File: subsys/net/ip/tcp.c [hunk 1]
```diff
@@ -2193,6 +2193,10 @@ int net_tcp_queue_data(struct net_context *context, struct net_pkt *pkt)
 	k_mutex_lock(&conn->lock, K_FOREVER);
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:2202]: Minor typo in commit `net: tcp: Simple zero window probe implementation`
s/Proble/Probe/

File: subsys/net/ip/tcp.c [hunk 2]
```diff
@@ -51,6 +51,7 @@ static struct k_work_q tcp_work_q;
 static K_KERNEL_STACK_DEFINE(work_q_stack, CONFIG_NET_TCP_WORKQ_STACK_SIZE);
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:51]: Typo in commit `net: tcp: Retrigger the send work in case of loopback`
s/exectue/execute/

**Review by jukkar — DISMISSED** [review]
Looks good, just minor nits in documentation.

**Comment by rlubos** [comment]
@jukkar Thanks, I've updated the commit descriptions.

**Review by jukkar — APPROVED** [review]

**Comment by ssharks** [comment]
The socket in the testcase is blocking. Nevertheless the send command returns without having the complete amount of data transferred. This is resolved by calling send calls until all the data has been transferred. Is this in line with the POSIX definition of send calls?

From: https://man7.org/linux/man-pages/man2/send.2.html
`When the message does not fit into the send buffer of the socket,
 send() normally blocks, unless the socket has been placed in
 nonblocking I/O mode.  In nonblocking mode it would fail with the error EAGAIN or EWOULDBLOCK in this case.  The [select(2)](https://man7.org/linux/man-pages/man2/select.2.html) call may be used to determine when it is possible to send more data.`

**Comment by rlubos** [comment]
> The socket in the testcase is blocking. Nevertheless the send command returns without having the complete amount of data transferred. This is resolved by calling send calls until all the data has been transferred. Is this in line with the POSIX definition of send calls?
> 
> From: https://man7.org/linux/man-pages/man2/send.2.html `When the message does not fit into the send buffer of the socket, send() normally blocks, unless the socket has been placed in nonblocking I/O mode. In nonblocking mode it would fail with the error EAGAIN or EWOULDBLOCK in this case. The [select(2)](https://man7.org/linux/man-pages/man2/select.2.html) call may be used to determine when it is possible to send more data.`

Yes, for STREAM socket it's ok to send fewer bytes than requested. It's not explicitly stated in the man page you quote, but it says that `send()` is equivalent of `write()`, which makes makes it clear:
>  Note that a successful write() may transfer fewer than count bytes.

**Comment by ssharks** [comment]
@rlubos Thanks for your explanation on the expected behavior of send.

I applied your changes on top of NRF Connect SDK 1.9.1 and validated that your changes work on my embedded system.

**Review by carlescufi — APPROVED** [review]

**Pull request merged**